src/query/builder.js

```diff
@@ -68,6 +68,44 @@ assign(Builder.prototype, {
     return this;
   },
 
+  // With
+  // ------
+
+  with(alias, statement, bindings) {
+    if (typeof statement === 'function') {
+      return this.withWrapped(alias, statement);
+    }
+    // Allow a raw statement to be passed along to the query.
+    if (statement instanceof Raw && arguments.length >= 2) {
+      return this.withRaw(alias, statement, bindings);
+    }
+
```

Comment:
I suppose that this line should never be run, since only allowed parameters are function and knex.raw. Throwing exception and adding test for that would be good to prevent `undefined does not have property ...` errors.